File: .setuptools-cmake-build/_deps/pybind11-src/tests/test_type_caster_pyobject_ptr.py
from __future__ import annotations

import pytest

from pybind11_tests import type_caster_pyobject_ptr as m


# For use as a temporary user-defined object, to maximize sensitivity of the tests below.
class ValueHolder:
    def __init__(self, value):
        self.value = value


def test_cast_from_pyobject_ptr():
    assert m.cast_from_pyobject_ptr() == 6758


def test_cast_handle_to_pyobject_ptr():
    assert m.cast_handle_to_pyobject_ptr(ValueHolder(24)) == 76


def test_cast_object_to_pyobject_ptr():
    assert m.cast_object_to_pyobject_ptr(ValueHolder(43)) == 257


def test_cast_list_to_pyobject_ptr():
    assert m.cast_list_to_pyobject_ptr([1, 2, 3, 4, 5]) == 395


def test_return_pyobject_ptr():
    assert m.return_pyobject_ptr() == 2314


def test_pass_pyobject_ptr():
    assert m.pass_pyobject_ptr(ValueHolder(82)) == 118


@pytest.mark.parametrize(
    "call_callback",
    [
        m.call_callback_with_object_return,
        m.call_callback_with_pyobject_ptr_return,
    ],
)
def test_call_callback_with_object_return(call_callback):
    def cb(value):
        if value < 0:
            raise ValueError("Raised from cb")
        return ValueHolder(1000 - value)

    assert call_callback(cb, 287).value == 713

    with pytest.raises(ValueError, match="^Raised from cb$"):
        call_callback(cb, -1)


def test_call_callback_with_pyobject_ptr_arg():
    def cb(obj):
        return 300 - obj.value

    assert m.call_callback_with_pyobject_ptr_arg(cb, ValueHolder(39)) == 261


@pytest.mark.parametrize("set_error", [True, False])
def test_cast_to_python_nullptr(set_error):
    expected = {
        True: r"^Reflective of healthy error handling\.$",
        False: (
            r"^Internal error: pybind11::error_already_set called "
            r"while Python error indicator not set\.$"
        ),
    }[set_error]
    with pytest.raises(RuntimeError, match=expected):
        m.cast_to_pyobject_ptr_nullptr(set_error)


def test_cast_to_python_non_nullptr_with_error_set():
    with pytest.raises(SystemError) as excinfo:
        m.cast_to_pyobject_ptr_non_nullptr_with_error_set()
    assert str(excinfo.value) == "src != nullptr but PyErr_Occurred()"
    assert str(excinfo.value.__cause__) == "Reflective of unhealthy error handling."


def test_pass_list_pyobject_ptr():
    acc = m.pass_list_pyobject_ptr([ValueHolder(842), ValueHolder(452)])
    assert acc == 842452


def test_return_list_pyobject_ptr_take_ownership():
    vec_obj = m.return_list_pyobject_ptr_take_ownership(ValueHolder)
    assert [e.value for e in vec_obj] == [93, 186]


def test_return_list_pyobject_ptr_reference():
    vec_obj = m.return_list_pyobject_ptr_reference(ValueHolder)
    assert [e.value for e in vec_obj] == [93, 186]
    # Commenting out the next `assert` will leak the Python references.
    # An easy way to see evidence of the leaks:
    # Insert `while True:` as the first line of this function and monitor the
    # process RES (Resident Memory Size) with the Unix top command.
    assert m.dec_ref_each_pyobject_ptr(vec_obj) == 2


def test_type_caster_name_via_incompatible_function_arguments_type_error():
    with pytest.raises(TypeError, match=r"1\. \(arg0: object, arg1: int\) -> None"):
        m.pass_pyobject_ptr_and_int(ValueHolder(101), ValueHolder(202))


def test_trampoline_with_pyobject_ptr_return():
    class Drvd(m.WithPyObjectPtrReturn):
        def return_pyobject_ptr(self):
            return ["11", "22", "33"]

    # Basic health check: First make sure this works as expected.
    d = Drvd()
    assert d.return_pyobject_ptr() == ["11", "22", "33"]

    while True:
        # This failed before PR #5156: AddressSanitizer: heap-use-after-free ... in Py_DECREF
        d_repr = m.call_return_pyobject_ptr(d)
        assert d_repr == repr(["11", "22", "33"])
        break  # Comment out for manual leak checking.

File: .setuptools-cmake-build/_deps/pybind11-src/tests/test_type_caster_pyobject_ptr.cpp
#include <pybind11/functional.h>
#include <pybind11/stl.h>
#include <pybind11/type_caster_pyobject_ptr.h>

#include "pybind11_tests.h"

#include <cstddef>
#include <string>
#include <vector>

namespace test_type_caster_pyobject_ptr {

std::vector<PyObject *> make_vector_pyobject_ptr(const py::object &ValueHolder) {
    std::vector<PyObject *> vec_obj;
    for (int i = 1; i < 3; i++) {
        vec_obj.push_back(ValueHolder(i * 93).release().ptr());
    }
    // This vector now owns the refcounts.
    return vec_obj;
}

struct WithPyObjectPtrReturn {
#if defined(__clang_major__) && __clang_major__ < 4
    WithPyObjectPtrReturn() = default;
    WithPyObjectPtrReturn(const WithPyObjectPtrReturn &) = default;
#endif
    virtual ~WithPyObjectPtrReturn() = default;
    virtual PyObject *return_pyobject_ptr() const = 0;
};

struct WithPyObjectPtrReturnTrampoline : WithPyObjectPtrReturn {
    PyObject *return_pyobject_ptr() const override {
        PYBIND11_OVERRIDE_PURE(PyObject *, WithPyObjectPtrReturn, return_pyobject_ptr,
                               /* no arguments */);
    }
};

std::string call_return_pyobject_ptr(const WithPyObjectPtrReturn *base_class_ptr) {
    PyObject *returned_obj = base_class_ptr->return_pyobject_ptr();
#if !defined(PYPY_VERSION) // It is not worth the trouble doing something special for PyPy.
    if (Py_REFCNT(returned_obj) != 1) {
        py::pybind11_fail(__FILE__ ":" PYBIND11_TOSTRING(__LINE__));
    }
#endif
    auto ret_val = py::repr(returned_obj).cast<std::string>();
    Py_DECREF(returned_obj);
    return ret_val;
}

} // namespace test_type_caster_pyobject_ptr

TEST_SUBMODULE(type_caster_pyobject_ptr, m) {
    using namespace test_type_caster_pyobject_ptr;

    m.def("cast_from_pyobject_ptr", []() {
        PyObject *ptr = PyLong_FromLongLong(6758L);
        return py::cast(ptr, py::return_value_policy::take_ownership);
    });
    m.def("cast_handle_to_pyobject_ptr", [](py::handle obj) {
        auto rc1 = obj.ref_count();
        auto *ptr = py::cast<PyObject *>(obj);
        auto rc2 = obj.ref_count();
        if (rc2 != rc1 + 1) {
            return -1;
        }
        return 100 - py::reinterpret_steal<py::object>(ptr).attr("value").cast<int>();
    });
    m.def("cast_object_to_pyobject_ptr", [](py::object obj) {
        py::handle hdl = obj;
        auto rc1 = hdl.ref_count();
        auto *ptr = py::cast<PyObject *>(std::move(obj));
        auto rc2 = hdl.ref_count();
        if (rc2 != rc1) {
            return -1;
        }
        return 300 - py::reinterpret_steal<py::object>(ptr).attr("value").cast<int>();
    });
    m.def("cast_list_to_pyobject_ptr", [](py::list lst) {
        // This is to cover types implicitly convertible to object.
        py::handle hdl = lst;
        auto rc1 = hdl.ref_count();
        auto *ptr = py::cast<PyObject *>(std::move(lst));
        auto rc2 = hdl.ref_count();
        if (rc2 != rc1) {
            return -1;
        }
        return 400 - static_cast<int>(py::len(py::reinterpret_steal<py::list>(ptr)));
    });

    m.def(
        "return_pyobject_ptr",
        []() { return PyLong_FromLongLong(2314L); },
        py::return_value_policy::take_ownership);
    m.def("pass_pyobject_ptr", [](PyObject *ptr) {
        return 200 - py::reinterpret_borrow<py::object>(ptr).attr("value").cast<int>();
    });

    m.def("call_callback_with_object_return",
          [](const std::function<py::object(int)> &cb, int value) { return cb(value); });
    m.def(
        "call_callback_with_pyobject_ptr_return",
        [](const std::function<PyObject *(int)> &cb, int value) { return cb(value); },
        py::return_value_policy::take_ownership);
    m.def(
        "call_callback_with_pyobject_ptr_arg",
        [](const std::function<int(PyObject *)> &cb, py::handle obj) { return cb(obj.ptr()); },
        py::arg("cb"), // This triggers return_value_policy::automatic_reference
        py::arg("obj"));

    m.def("cast_to_pyobject_ptr_nullptr", [](bool set_error) {
        if (set_error) {
            py::set_error(PyExc_RuntimeError, "Reflective of healthy error handling.");
        }
        PyObject *ptr = nullptr;
        py::cast(ptr);
    });

    m.def("cast_to_pyobject_ptr_non_nullptr_with_error_set", []() {
        py::set_error(PyExc_RuntimeError, "Reflective of unhealthy error handling.");
        py::cast(Py_None);
    });

    m.def("pass_list_pyobject_ptr", [](const std::vector<PyObject *> &vec_obj) {
        int acc = 0;
        for (const auto &ptr : vec_obj) {
            acc = acc * 1000 + py::reinterpret_borrow<py::object>(ptr).attr("value").cast<int>();
        }
        return acc;
    });

    m.def("return_list_pyobject_ptr_take_ownership",
          make_vector_pyobject_ptr,
          // Ownership is transferred one-by-one when the vector is converted to a Python list.
          py::return_value_policy::take_ownership);

    m.def("return_list_pyobject_ptr_reference",
          make_vector_pyobject_ptr,
          // Ownership is not transferred.
          py::return_value_policy::reference);

    m.def("dec_ref_each_pyobject_ptr", [](const std::vector<PyObject *> &vec_obj) {
        std::size_t i = 0;
        for (; i < vec_obj.size(); i++) {
            py::handle h(vec_obj[i]);
            if (static_cast<std::size_t>(h.ref_count()) < 2) {
                break; // Something is badly wrong.
            }
            h.dec_ref();
        }
        return i;
    });

    m.def("pass_pyobject_ptr_and_int", [](PyObject *, int) {});

#ifdef PYBIND11_NO_COMPILE_SECTION // Change to ifndef for manual testing.
    {
        PyObject *ptr = nullptr;
        (void) py::cast(*ptr);
    }
#endif

    py::class_<WithPyObjectPtrReturn, WithPyObjectPtrReturnTrampoline>(m, "WithPyObjectPtrReturn")
        .def(py::init<>())
        .def("return_pyobject_ptr", &WithPyObjectPtrReturn::return_pyobject_ptr);

    m.def("call_return_pyobject_ptr", call_return_pyobject_ptr);
}
